apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/KmsTest.cs

```diff
@@ -0,0 +1,251 @@
+            Assert.StartsWith(VersionedKeyName1Prefix, obj.KmsKeyName);
+
+            // We should be able to download it again, even with a default (no CSEK) client.
+            TestHelpers.ValidateData(bucketName, objectName, data);
+        }
+
+        [SkippableFact]
+        public void UploadObject_BucketHasDefaultKmsKey_UploadWithNoSpecifiedKey_DownloadWithClientWithCsek()
+        {
+            Skip.If(SkipTests);
+            var client = _fixture.Client;
+
+            string bucketName = CreateBucketWithDefaultKmsKeyName(KeyName1);
+            string objectName = TestHelpers.GenerateName();
+            var data = TestHelpers.GenerateData(1024);
+
+            // Upload an object without specifying anything, so it should be encrypted with key 1.
+            Object obj = client.UploadObject(bucketName, objectName, null, data);
+            Assert.StartsWith(VersionedKeyName1Prefix, obj.KmsKeyName);
+
+            // We should be able to download the file even if we use a client that supplies CSEK by default.
+            EncryptionKey key = EncryptionKey.Generate();
+            var encryptionKeyClient = StorageClient.Create(encryptionKey: key);
+            TestHelpers.ValidateData(bucketName, objectName, data, encryptionKeyClient);
+        }
+
+        [SkippableFact]
+        public void UploadObject_SpecifyBothCsekAndKmsKey()
+        {
+            Skip.If(SkipTests);
+            var client = _fixture.Client;
+
+            string objectName = TestHelpers.GenerateName();
+            var data = TestHelpers.GenerateData(1024);
+
+            // It's invalid to specify both an EncryptionKey (that's non-None) and a key name
+            var options = new UploadObjectOptions { EncryptionKey = EncryptionKey.Generate(), KmsKeyName = KeyName1 };
+            var exception = Assert.Throws<ArgumentException>(() => client.UploadObject(_fixture.InitiallyEmptyBucket, objectName, null, data, options));
+            Assert.Contains(nameof(CopyObjectOptions.EncryptionKey), exception.Message);
+            Assert.Contains(nameof(CopyObjectOptions.KmsKeyName), exception.Message);
+        }
+
+        [SkippableFact]
+        public void CopyObject_ChangeKey()
+        {
+            Skip.If(SkipTests);
+            var client = _fixture.Client;
+            string bucketName = _fixture.InitiallyEmptyBucket;
+            _fixture.CreateBucket(bucketName, false);
+            string objectName = TestHelpers.GenerateName();
+            var data = TestHelpers.GenerateData(1024);
+
+            // Upload an object with a KMS key.
+            Object obj = client.UploadObject(bucketName, objectName, null, data, new UploadObjectOptions { KmsKeyName = KeyName1 });
+            Assert.StartsWith(VersionedKeyName1Prefix, obj.KmsKeyName);
+
+            // Copy it (within the same bucket, for simplicity) with a different key
```

Comment:
Hi @jskeet, I'm looking for the test which exercises migration from CSEK to CMEK and I didn't seem to find it. Does this test exist?

Reply:
Good catch - I only had one from one CMEK to another. I've added a test from CSEK to CMEK. (And it works :)